benchmark/bench/server-stress.js

```diff
@@ -0,0 +1,82 @@
+import fs from 'fs/promises';
+import { fileURLToPath } from 'url';
+import autocannon from 'autocannon';
+import { execaCommand } from 'execa';
+import { waitUntilBusy } from 'port-authority';
+import { astroBin } from './_util.js';
+
+const port = 4321;
+
+export const defaultProject = 'server-stress-default';
+
+/**
+ * @param {URL} projectDir
+ * @param {URL} outputFile
+ */
+export async function run(projectDir, outputFile) {
+	const root = fileURLToPath(projectDir);
+
+	console.log('Building...');
+	await execaCommand(`${astroBin} build`, {
+		cwd: root,
+		stdio: 'inherit',
+	});
+
+	console.log('Previewing...');
+	const previewProcess = execaCommand(`${astroBin} preview --port ${port}`, {
+		cwd: root,
+		stdio: 'inherit',
+	});
+
+	console.log('Waiting for server ready...');
+	await waitUntilBusy(port, { timeout: 5000 });
+
+	console.log('Running benchmark...');
+	const result = await benchmarkCannon();
+
+	console.log('Killing server...');
+	if (!previewProcess.kill('SIGTERM')) {
+		console.warn('Failed to kill server process id:', previewProcess.pid);
+	}
+
+	console.log('Writing results to', fileURLToPath(outputFile));
+	await fs.writeFile(outputFile, JSON.stringify(result, null, 2));
+
+	console.log('Result preview:');
+	console.log('='.repeat(10));
+	console.log(`#### Server stress\n\n`);
+	let text = autocannon.printResult(result);
+	if (process.env.CI) {
+		text = text.match(/^.*?requests in.*?read$/m)?.[0];
```

Comment:
Would you mind adding a small comment explaining what we're doing here and why?

Reply:
Good idea 👍 It's here mostly to truncate the markdown text in the comments so it's not huge. I'll update this.